app/models/host/base.rb

```diff
@@ -220,15 +220,14 @@ def set_comment(parser)
       example '@host.facts["uptime"] # => "30 days"', desc: 'Getting specific fact value, +uptime+ in this case'
       aliases :facts
     end
-    def facts_hash(fact_names = [], *args)
-      filtered_names = [fact_names, *args].flatten.compact
-      hash = {}
-      query = fact_values.includes(:fact_name)
-      query = query.where(fact_names: {name: filtered_names}) if filtered_names.present?
-      query.each do |fact|
-        hash[fact.fact_name.name] = fact.value
-      end
-      hash
+    def facts_hash(*fact_names)
+      # keep it to one SQL query
+      query = if fact_names.present?
+                fact_values.includes(:fact_name).where(fact_names: {name: fact_names})
+              else
+                fact_values.includes(:fact_name)
+              end
```

Comment:
Does it really trigger a query if you use:
```suggestion
      query = fact_values.includes(:fact_name)
      query = query.where(fact_names: {name: fact_names}) if fact_names.present?
```

I thought we used that query builder pattern quite a lot and it didn't trigger additional queries.

Reply:
Yes, I confirmed it in the Rails console:

```
query = rhel9.fact_values.includes(:fact_name)
2023-09-18T15:45:40 [D|sql|]   FactValue Load (0.5ms)  SELECT "fact_values".* FROM "fact_values" WHERE "fact_values"."host_id" = $1  [["host_id", 3]]
2023-09-18T15:45:40 [D|sql|]   FactName Load (1.1ms)  SELECT "fact_names".* FROM "fact_names" WHERE "fact_names"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67, $68, $69, $70, $71, $72, $73, $74, $75, $76, $77, $78, $79, $80, $81, $82, $83, $84, $85, $86, $87, $88, $89, $90, $91, $92, $93, $94, $95, $96, $97, $98, $99, $100, $101, $102, $103, $104, $105, $106, $107, $108, $109, $110, $111, $112, $113, $114, $115, $116, $117, $118, $119, $120, $121, $122, $123, $124, $125, $126, $127, $128, $129, $130, $131, $132, $133, $134, $135, $136, $137, $138, $139, $140, $141, $142, $143, $144, $145, $146, $147, $148, $149, $150, $151, $152, $153, $154, $155, $156, $157, $158, $159, $160, $161, $162, $163, $164, $165, $166, $167, $168, $169, $170, $171, $172, $173, $174) ORDER BY fact_names.name

query = fact_names.present? ? rhel9.fact_values.includes(:fact_name).where(fact_names: {name: fact_names}) : rhel9.fact_values.includes(:fact_name)
2023-09-18T15:47:33 [D|sql|]   SQL (0.8ms)  SELECT "fact_values"."id" AS t0_r0, "fact_values"."value" AS t0_r1, "fact_values"."fact_name_id" AS t0_r2, "fact_values"."host_id" AS t0_r3, "fact_values"."updated_at" AS t0_r4, "fact_values"."created_at" AS t0_r5, "fact_names"."id" AS t1_r0, "fact_names"."name" AS t1_r1, "fact_names"."updated_at" AS t1_r2, "fact_names"."created_at" AS t1_r3, "fact_names"."compose" AS t1_r4, "fact_names"."short_name" AS t1_r5, "fact_names"."type" AS t1_r6, "fact_names"."ancestry" AS t1_r7 FROM "fact_values" LEFT OUTER JOIN "fact_names" ON "fact_names"."id" = "fact_values"."fact_name_id" WHERE "fact_values"."host_id" = $1 AND "fact_names"."name" IN ($2, $3)
```

Reply:
Now that I'm looking at it further: why do we even need `includes`? Now that we're using `pluck()` I think it's redundant.

Reply:
Seems it's needed:

![image](https://github.com/theforeman/foreman/assets/22042343/b1c89bd7-813a-4efe-9fa7-28edc0492304)

Reply:
Isn't `query.joins(:fact_name)` the right way to go here?

Reply:
That seems to work :+1: 

Updated